src/typecheck/mod.rs

```diff
@@ -2237,6 +2228,7 @@ pub fn unify(
                     body: body2,
                 },
             ) if var_kind1 == var_kind2 => {
+                // XXX: ^^ do we want to compare std::mem::discriminant() or also check equality of `excluded`s above?
```

Comment:
At first sight, I would say that the discriminant is sufficient, because equality of the substituted types as checked below probably implies that the discriminant are the same.

On the other hand, we fail early and avoid any traversal or substitution if we can determine that the discriminants aren't the same directly :man_shrugging: . In the end it depends on what is the "happy path" and what we want to optimize for. I would say we want to optimize for when the type are equal, because otherwise we fail anyway, and traversing an additional type won't change much (you'll get your error message 0.010s later). On the other hand, if for some reason a lot of polymorphic types are unified, then you'll pay the price of checking the additional discriminants for each such unification.

Honestly, in any case, this check is dumbly slow (as per the comment), and shouldn't fire a lot (ideally, most of the unifications involving polymorphic types happen when types are already instantiated). It's probably fine either way.

Reply:
Oh, yes, thank you. I discussed this with Joe and we came to basically the same conclusion. I'd meant to take out the comment but forgot. Good to get confirmation anyways.